File: src/lib/projectTree.ts
import type { Project } from "./types";

/** Returns the direct children of `parentId` (or every top-level project, if `parentId` is `undefined`), in the order they appear in `projects`. */
export function childrenOf(projects: Project[], parentId: string | undefined): Project[] {
  return projects.filter((p) => p.parent_id === parentId);
}

/** Returns the ancestors of the project identified by `id`, nearest-first, ending at the root. Empty if `id` doesn't exist in `projects` or names a top-level project. */
export function ancestorsOf(projects: Project[], id: string): Project[] {
  const result: Project[] = [];
  let currentId = projects.find((p) => p.id === id)?.parent_id;

  while (currentId !== undefined) {
    const ancestor = projects.find((p) => p.id === currentId);
    if (!ancestor) break;
    result.push(ancestor);
    currentId = ancestor.parent_id;
  }

  return result;
}

/** Returns the project identified by `id` (if found) followed by its ancestors, nearest-first — the full settings-resolution chain for that project, ending at the root. Empty if `id` doesn't exist in `projects`. */
export function selfAndAncestors(projects: Project[], id: string): Project[] {
  const self = projects.find((p) => p.id === id);
  if (!self) return [];
  return [self, ...ancestorsOf(projects, id)];
}

/** Returns the full ancestor path for the project identified by `id`, root-first and joined with "/" (e.g. "Work/Client A/Phase 1") — unambiguous even for same-named subprojects under different parents, unlike a bare leaf name. Empty string if `id` doesn't exist in `projects`. */
export function projectPath(projects: Project[], id: string): string {
  const self = projects.find((p) => p.id === id);
  if (!self) return "";
  const ancestorsRootFirst = ancestorsOf(projects, id).reverse();
  return [...ancestorsRootFirst, self].map((p) => p.name).join("/");
}

/**
 * Formats root-first path `segments` (e.g. from `projectPath`, split back
 * apart) as a literal `+Project` quick-add token's text — the inverse of
 * `parsePathSegments` in `naturalLanguage.ts` — quoting any segment that
 * contains whitespace so the result round-trips back through the parser
 * unambiguously (e.g. `["Work", "Client A"]` becomes `Work/"Client A"`).
 */
export function formatProjectPathToken(segments: string[]): string {
  return segments.map((segment) => (/\s/.test(segment) ? `"${segment}"` : segment)).join("/");
}

/**
 * Resolves a root-first ancestor path (e.g. `["Work", "Client A"]`, the
 * counterpart to `projectPath`'s string form) to the project it names: at
 * each level, the first child (or, for the first segment, the first
 * top-level project) whose name matches case-insensitively — mirroring
 * `find_project`'s existing case-insensitivity. Returns `undefined` as soon
 * as any level has no match (an invalid/typo'd path doesn't fall back to
 * matching the last segment as a bare name elsewhere in the tree — that
 * would be more confusing than just not matching), or if `segments` is
 * empty.
 */
export function findProjectByPath(projects: Project[], segments: string[]): Project | undefined {
  if (segments.length === 0) return undefined;

  let parentId: string | undefined = undefined;
  let match: Project | undefined;

  for (const segment of segments) {
    const lowerSegment = segment.toLowerCase();
    match = childrenOf(projects, parentId).find((p) => p.name.toLowerCase() === lowerSegment);
    if (!match) return undefined;
    parentId = match.id;
  }

  return match;
}

/** Returns every transitive descendant of the project identified by `id` (children, grandchildren, ...). */
export function descendantsOf(projects: Project[], id: string): Project[] {
  const result: Project[] = [];
  const frontier: string[] = [id];

  while (frontier.length > 0) {
    const currentId = frontier.pop();
    if (currentId === undefined) break;
    for (const child of childrenOf(projects, currentId)) {
      result.push(child);
      frontier.push(child.id);
    }
  }

  return result;
}

/** Returns `true` if making `newParentId` the parent of `movingId` would create a cycle — i.e. `newParentId` is `movingId` itself, or is one of `movingId`'s current descendants. */
export function wouldCreateCycle(projects: Project[], movingId: string, newParentId: string): boolean {
  if (movingId === newParentId) return true;
  return descendantsOf(projects, movingId).some((p) => p.id === newParentId);
}

export interface ProjectOrderUpdate {
  id: string;
  parent_id: string | undefined;
  order: number;
}

/**
 * Computes the project updates needed to persist `zoneItems` (a project
 * tree zone's current children, in display order, after a drag-and-drop
 * reorder/reparent) as the children of `zoneParentId` (`undefined` for the
 * top-level zone). Returns one `ProjectOrderUpdate` per item whose
 * `parent_id` and/or `order` actually needs to change — an item already
 * correctly parented and numbered isn't included, so callers only persist
 * what actually changed.
 *
 * `allProjects` (the full, current tree) is used for a cycle check: if any
 * item's `parent_id` would need to change to `zoneParentId`, but
 * `zoneParentId` is that item's own descendant (see `wouldCreateCycle`),
 * the whole result is `rejected: true` and `updates` is empty — callers
 * should treat a rejected drop as entirely invalid and re-sync from the
 * server rather than partially applying it.
 */
export function computeZoneOrderUpdates(
  allProjects: Project[],
  zoneParentId: string | undefined,
  zoneItems: Project[],
  orderStep = 1000,
): { updates: ProjectOrderUpdate[]; rejected: boolean } {
  const updates: ProjectOrderUpdate[] = [];

  for (const [index, item] of zoneItems.entries()) {
    const needsReparent = item.parent_id !== zoneParentId;
    if (needsReparent && zoneParentId !== undefined && wouldCreateCycle(allProjects, item.id, zoneParentId)) {
      return { updates: [], rejected: true };
    }

    const newOrder = (index + 1) * orderStep;
    if (needsReparent || item.order !== newOrder) {
      updates.push({ id: item.id, parent_id: zoneParentId, order: newOrder });
    }
  }

  return { updates, rejected: false };
}

File: src-tauri/src/project_tree.rs
use crate::project::Project;

/// Returns the direct children of `parent_id` (or every top-level project,
/// if `parent_id` is `None`), in the order they appear in `projects`.
#[allow(dead_code)]
pub fn children_of<'a>(projects: &'a [Project], parent_id: Option<&str>) -> Vec<&'a Project> {
    projects
        .iter()
        .filter(|p| p.parent_id.as_deref() == parent_id)
        .collect()
}

/// Returns the ancestors of the project identified by `id`, nearest-first,
/// ending at the root. Empty if `id` doesn't exist in `projects` or names a
/// top-level project.
#[allow(dead_code)]
pub fn ancestors_of<'a>(projects: &'a [Project], id: &str) -> Vec<&'a Project> {
    let mut result = Vec::new();
    let mut current_id = projects
        .iter()
        .find(|p| p.id == id)
        .and_then(|p| p.parent_id.clone());

    while let Some(ancestor_id) = current_id {
        let Some(ancestor) = projects.iter().find(|p| p.id == ancestor_id) else {
            break;
        };
        result.push(ancestor);
        current_id = ancestor.parent_id.clone();
    }

    result
}

/// Returns the project identified by `id` (if found) followed by its
/// ancestors, nearest-first — the full settings-resolution chain for that
/// project, ending at the root. Empty if `id` doesn't exist in `projects`.
#[allow(dead_code)]
pub fn self_and_ancestors<'a>(projects: &'a [Project], id: &str) -> Vec<&'a Project> {
    let mut result: Vec<&Project> = projects.iter().find(|p| p.id == id).into_iter().collect();
    result.extend(ancestors_of(projects, id));
    result
}

/// Returns every transitive descendant of the project identified by `id`
/// (children, grandchildren, ...). Order is not guaranteed to be any
/// particular traversal order — callers needing a specific order should sort
/// the result themselves.
#[allow(dead_code)]
pub fn descendants_of<'a>(projects: &'a [Project], id: &str) -> Vec<&'a Project> {
    let mut result: Vec<&Project> = Vec::new();
    let mut frontier: Vec<String> = vec![id.to_string()];

    while let Some(current_id) = frontier.pop() {
        for child in children_of(projects, Some(current_id.as_str())) {
            result.push(child);
            frontier.push(child.id.clone());
        }
    }

    result
}

/// Returns `true` if making `new_parent_id` the parent of `moving_id` would
/// create a cycle — i.e. `new_parent_id` is `moving_id` itself, or is one of
/// `moving_id`'s current descendants.
#[allow(dead_code)]
pub fn would_create_cycle(projects: &[Project], moving_id: &str, new_parent_id: &str) -> bool {
    if moving_id == new_parent_id {
        return true;
    }
    descendants_of(projects, moving_id)
        .iter()
        .any(|p| p.id == new_parent_id)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Builds a small fixture tree:
    /// ```text
    /// root_a (top-level)
    /// ├── child_a1
    /// │   └── grandchild_a1a
    /// └── child_a2
    /// root_b (top-level, no children)
    /// ```
    fn fixture_tree() -> Vec<Project> {
        let mut root_a = Project::new("Root A".to_string(), "#111111".to_string(), 1);
        root_a.id = "root_a".to_string();

        let mut root_b = Project::new("Root B".to_string(), "#222222".to_string(), 2);
        root_b.id = "root_b".to_string();

        let mut child_a1 = Project::new("Child A1".to_string(), "#333333".to_string(), 1);
        child_a1.id = "child_a1".to_string();
        child_a1.parent_id = Some("root_a".to_string());

        let mut child_a2 = Project::new("Child A2".to_string(), "#444444".to_string(), 2);
        child_a2.id = "child_a2".to_string();
        child_a2.parent_id = Some("root_a".to_string());

        let mut grandchild_a1a =
            Project::new("Grandchild A1a".to_string(), "#555555".to_string(), 1);
        grandchild_a1a.id = "grandchild_a1a".to_string();
        grandchild_a1a.parent_id = Some("child_a1".to_string());

        vec![root_a, root_b, child_a1, child_a2, grandchild_a1a]
    }

    #[test]
    fn children_of_none_returns_top_level_projects() {
        let projects = fixture_tree();

        let children = children_of(&projects, None);

        let ids: Vec<&str> = children.iter().map(|p| p.id.as_str()).collect();
        assert_eq!(ids, vec!["root_a", "root_b"]);
    }

    #[test]
    fn children_of_returns_direct_children_only() {
        let projects = fixture_tree();

        let children = children_of(&projects, Some("root_a"));

        let ids: Vec<&str> = children.iter().map(|p| p.id.as_str()).collect();
        assert_eq!(ids, vec!["child_a1", "child_a2"]);
    }

    #[test]
    fn children_of_returns_empty_for_a_leaf() {
        let projects = fixture_tree();

        let children = children_of(&projects, Some("child_a2"));

        assert!(children.is_empty());
    }

    #[test]
    fn ancestors_of_top_level_project_is_empty() {
        let projects = fixture_tree();

        let ancestors = ancestors_of(&projects, "root_a");

        assert!(ancestors.is_empty());
    }

    #[test]
    fn ancestors_of_returns_nearest_first() {
        let projects = fixture_tree();

        let ancestors = ancestors_of(&projects, "grandchild_a1a");

        let ids: Vec<&str> = ancestors.iter().map(|p| p.id.as_str()).collect();
        assert_eq!(ids, vec!["child_a1", "root_a"]);
    }

    #[test]
    fn ancestors_of_missing_id_is_empty() {
        let projects = fixture_tree();

        let ancestors = ancestors_of(&projects, "does_not_exist");

        assert!(ancestors.is_empty());
    }

    #[test]
    fn ancestors_of_stops_at_a_dangling_parent_id() {
        let mut projects = fixture_tree();
        // Simulate a parent that was deleted without cleaning up this
        // reference: child_a1's parent_id points at a project that no
        // longer exists.
        projects[2].parent_id = Some("deleted_project".to_string());

        let ancestors = ancestors_of(&projects, "grandchild_a1a");

        let ids: Vec<&str> = ancestors.iter().map(|p| p.id.as_str()).collect();
        assert_eq!(ids, vec!["child_a1"]);
    }

    #[test]
    fn self_and_ancestors_includes_self_first() {
        let projects = fixture_tree();

        let chain = self_and_ancestors(&projects, "grandchild_a1a");

        let ids: Vec<&str> = chain.iter().map(|p| p.id.as_str()).collect();
        assert_eq!(ids, vec!["grandchild_a1a", "child_a1", "root_a"]);
    }

    #[test]
    fn self_and_ancestors_for_missing_id_is_empty() {
        let projects = fixture_tree();

        let chain = self_and_ancestors(&projects, "does_not_exist");

        assert!(chain.is_empty());
    }

    #[test]
    fn descendants_of_returns_all_levels() {
        let projects = fixture_tree();

        let descendants = descendants_of(&projects, "root_a");

        let mut ids: Vec<&str> = descendants.iter().map(|p| p.id.as_str()).collect();
        ids.sort_unstable();
        assert_eq!(ids, vec!["child_a1", "child_a2", "grandchild_a1a"]);
    }

    #[test]
    fn descendants_of_a_leaf_is_empty() {
        let projects = fixture_tree();

        let descendants = descendants_of(&projects, "grandchild_a1a");

        assert!(descendants.is_empty());
    }

    #[test]
    fn descendants_of_an_unrelated_root_is_empty() {
        let projects = fixture_tree();

        let descendants = descendants_of(&projects, "root_b");

        assert!(descendants.is_empty());
    }

    #[test]
    fn would_create_cycle_when_moving_under_self() {
        let projects = fixture_tree();

        assert!(would_create_cycle(&projects, "root_a", "root_a"));
    }

    #[test]
    fn would_create_cycle_when_moving_under_own_descendant() {
        let projects = fixture_tree();

        assert!(would_create_cycle(&projects, "root_a", "grandchild_a1a"));
    }

    #[test]
    fn would_not_create_cycle_when_moving_to_an_unrelated_project() {
        let projects = fixture_tree();

        assert!(!would_create_cycle(&projects, "child_a1", "root_b"));
    }

    #[test]
    fn would_not_create_cycle_when_moving_to_top_level() {
        let projects = fixture_tree();

        // Promoting child_a1 to top-level is modeled as re-parenting under
        // a synthetic id that isn't in the tree at all — would_create_cycle
        // only guards against id-based cycles, so this is just confirming
        // an arbitrary unrelated id is never flagged.
        assert!(!would_create_cycle(&projects, "child_a1", "unrelated_id"));
    }
}
